gdal/frmts/cderived/cderiveddataset.cpp

```diff
@@ -0,0 +1,196 @@
+/******************************************************************************
+ *
+ * Project:  GDAL
+ * Purpose:  Implementation of derived subdatasets
+ * Author:   Julien Michel <julien dot michel at cnes dot fr>
+ *
+ ******************************************************************************
+ * Copyright (c) 2008-2014 Antonio Valentino <antonio.valentino@tiscali.it>
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ *****************************************************************************/
+#include "../vrt/vrtdataset.h"
+#include "gdal_pam.h"
+#include "gdal_proxy.h"
+#include "derivedlist.h"
+
+#include <iostream>
+
+class ComplexDerivedDatasetContainer: public GDALPamDataset
+{
+    public:
+        ComplexDerivedDatasetContainer() {}
+};
+
+
+class ComplexDerivedDataset : public VRTDataset
+{
+  public:
+  ComplexDerivedDataset(int nXSize, int nYSize);
+  ~ComplexDerivedDataset();
+    
+    static GDALDataset *Open( GDALOpenInfo * );
+};
+
+
+ComplexDerivedDataset::ComplexDerivedDataset(int nXSize, int nYSize) : VRTDataset(nXSize,nYSize)
+{
+  poDriver = NULL;
+  SetWritable(FALSE);
+}
+
+ComplexDerivedDataset::~ComplexDerivedDataset()
+{
+}
+
+GDALDataset * ComplexDerivedDataset::Open(GDALOpenInfo * poOpenInfo)
+{
+  /* Try to open original dataset */
+  CPLString filename(poOpenInfo->pszFilename);
+ 
+  /* DERIVED_SUBDATASET should be first domain */
+  size_t dsds_pos = filename.find("DERIVED_SUBDATASET:");
+
+  if (dsds_pos == std::string::npos)
+    {
+    /* Unable to Open in this case */
+    return NULL;
+    }
+
+  /* Next, we need to now which derived dataset to compute */
+  size_t alg_pos = filename.find(":",dsds_pos+20);
+  if (alg_pos == std::string::npos)
+    {
+    /* Unable to Open if we do not find the name of the derived dataset */
+    return NULL;
+    }
+
+  CPLString odDerivedName = filename.substr(dsds_pos+19,alg_pos-dsds_pos-19);
```

Comment:
perhaps have a const size_t nPrefixLen = strlen("DERIVED_SUBDATASET:") to avoid magic number